File: src/07.greedy/01.arrayGreedy/17.406.ts
export function reconstructQueue(people: number[][]): number[][] {
    const n = people.length;

    // 排序, 按照升高 从大到小排序, 如果相同, 按照 k 从小到大排序
    people.sort((a, b) => {
        if (a[0] === b[0]) {
            return a[1] - b[1];
        }
        return b[0] - a[0];
    });

    const res: number[][] = [];
    for (let i = 0; i < n; i++) {
        // 将 该元素插入到 k 的位置
        res.splice(people[i][1], 0, people[i]);
    }

    return res;
}

File: src/07.greedy/01.arrayGreedy/17.406.go
package array_greedy

import "sort"

func ReconstructQueue(people [][]int) [][]int {
	// 排序
	sort.Slice(people, func(i, j int) bool {
		a := people[i]
		b := people[j]

		// 如果身高相同, 按照 k 小到大排序
		if a[0] == b[0] {
			return a[1] < b[1]
		}

		// 默认按照 身高从大到小排序
		return a[0] > b[0]
	})

	for i, v := range people {
		// copy(dest, src) 将 src 复制到 dest
		copy(people[v[1]+1:i+1], people[v[1]:i+1])
		people[v[1]] = v
	}

	return people
}
